File: libraries/typescript/src/providers/denoiser.ts
/**
 * Bring-your-own-license denoiser registry (TypeScript parity with the Python
 * `getpatter.providers.denoiser` module).
 *
 * Maps a small set of stable, string model ids to a resolved {@link AudioFilter}.
 * Selection is opt-in via `AgentOptions.denoiser`; the default (`undefined`)
 * leaves the inbound audio path unchanged.
 *
 * The Krisp models are **bring-your-own-license**: Patter ships ZERO Krisp
 * binaries, license keys, or `.kef` model files. Because Krisp does not publish
 * an official Node.js (server) SDK, resolving a `krisp-*` id constructs the
 * {@link KrispVivaFilter} scaffold, which throws the same clear guidance
 * (use the Python SDK, or provide a custom `AudioFilter`). This keeps the TS
 * API surface and error semantics identical to Python — a selected denoiser
 * never silently degrades to a no-op.
 */
import type { AgentOptions, AudioFilter } from '../types';
import { KrispModelKind, KrispVivaFilter } from './krisp-filter';

/** Env var pointing at the operator's directory of licensed Krisp `.kef` models. */
export const KRISP_MODELS_DIR_ENV = 'KRISP_MODELS_DIR';

/** Static description of a selectable denoiser model (parity with Python `DenoiserSpec`). */
export interface DenoiserSpec {
  /** Public, stable string id surfaced through `AgentOptions.denoiser`. */
  readonly modelId: string;
  /** Krisp session family the model loads through. */
  readonly kind: KrispModelKind;
  /** BYO `.kef` file expected under `KRISP_MODELS_DIR`. */
  readonly filename: string;
  /** Human-readable one-liner (docs / error context only). */
  readonly description: string;
}

/** The two REAL Krisp model ids Patter recognises. */
export const DENOISERS: Readonly<Record<string, DenoiserSpec>> = {
  'krisp-viva-tel-v2': {
    modelId: 'krisp-viva-tel-v2',
    kind: KrispModelKind.NC,
    filename: 'krisp-viva-tel-v2.kef',
    description: 'Krisp VIVA telephony noise cancellation (NC session).',
  },
  'krisp-bvc-o-pro-v3': {
    modelId: 'krisp-bvc-o-pro-v3',
    kind: KrispModelKind.BVC,
    filename: 'krisp-bvc-o-pro-v3.kef',
    description: 'Krisp Background Voice Cancellation, pro (BVC session).',
  },
} as const;

/** Registered denoiser ids (see {@link DENOISERS}). */
export const DENOISER_IDS: readonly string[] = Object.keys(DENOISERS);

/**
 * Resolve a string denoiser id to a concrete {@link AudioFilter}.
 *
 * @throws if `name` is not a registered denoiser id, or (for Krisp ids) the
 *   Node Krisp SDK is unavailable — the thrown error names the Python SDK /
 *   custom-AudioFilter alternatives.
 */
export function resolveDenoiser(name: string): AudioFilter {
  const spec = DENOISERS[name];
  if (!spec) {
    const known = [...DENOISER_IDS].sort().join(', ');
    throw new Error(`Unknown denoiser '${name}'. Known denoisers: ${known}.`);
  }
  const modelsDir = process.env[KRISP_MODELS_DIR_ENV];
  const modelPath = modelsDir ? `${modelsDir}/${spec.filename}` : undefined;
  // Constructs the scaffold, which throws the clear "no Node Krisp SDK" guidance
  // (single source of truth for the error). Kept structurally identical to
  // Python: the session family is passed through for parity.
  return new KrispVivaFilter({ modelPath, sessionType: spec.kind });
}

/**
 * Resolve the effective inbound audio filter for an agent.
 *
 * `audioFilter` (explicit instance) and `denoiser` (string id) are PARALLEL
 * opt-ins: when both are set the explicit `audioFilter` wins and the denoiser
 * is never resolved. Returns `undefined` when neither is set.
 */
export function resolveEffectiveAudioFilter(
  agent: Pick<AgentOptions, 'audioFilter' | 'denoiser'>,
): AudioFilter | undefined {
  if (agent.audioFilter) {
    return agent.audioFilter;
  }
  if (agent.denoiser) {
    return resolveDenoiser(agent.denoiser);
  }
  return undefined;
}

File: libraries/python/getpatter/providers/denoiser.py
"""Bring-your-own-license denoiser registry.

Maps a small set of stable, string model ids to a resolved
:class:`getpatter.providers.base.AudioFilter`. Selection is opt-in via
``Agent.denoiser``; the default (``None``) leaves the inbound audio path
byte-identical.

The Krisp models are **bring-your-own-license**: Patter ships ZERO Krisp
binaries, license keys, or ``.kef`` model files. Resolving a ``krisp-*`` id
constructs the existing :class:`~getpatter.providers.krisp_filter.KrispVivaFilter`
pointing at the operator's own model directory (``KRISP_MODELS_DIR`` + the
model id's ``.kef`` filename) and requires the operator's own ``krisp-audio``
SDK plus a ``KRISP_VIVA_SDK_LICENSE_KEY``. When the SDK, license, or model
file is missing, resolution raises a clear, actionable error naming exactly
what to install or set — it never silently degrades to a no-op filter.
"""

from __future__ import annotations

import os
from dataclasses import dataclass

from getpatter.providers.base import AudioFilter
from getpatter.providers.krisp_instance import KrispModelKind

# Environment variable pointing at the directory that holds the operator's
# licensed Krisp ``.kef`` model files (one per registered denoiser id).
KRISP_MODELS_DIR_ENV = "KRISP_MODELS_DIR"


@dataclass(frozen=True)
class DenoiserSpec:
    """Static description of a selectable denoiser model.

    Attributes
    ----------
    model_id:
        Public, stable string id surfaced through ``Agent.denoiser``.
    kind:
        Krisp session family the model loads through (:class:`KrispModelKind`).
    filename:
        BYO ``.kef`` file expected under ``KRISP_MODELS_DIR``.
    description:
        Human-readable one-liner (docs / error context only).
    """

    model_id: str
    kind: KrispModelKind
    filename: str
    description: str


# The two REAL Krisp model ids Patter recognises. Keep this list tight — each
# id must map to a concrete, documented Krisp model the operator can license.
_DENOISER_SPECS: tuple[DenoiserSpec, ...] = (
    DenoiserSpec(
        model_id="krisp-viva-tel-v2",
        kind=KrispModelKind.NC,
        filename="krisp-viva-tel-v2.kef",
        description="Krisp VIVA telephony noise cancellation (NC session).",
    ),
    DenoiserSpec(
        model_id="krisp-bvc-o-pro-v3",
        kind=KrispModelKind.BVC,
        filename="krisp-bvc-o-pro-v3.kef",
        description="Krisp Background Voice Cancellation, pro (BVC session).",
    ),
)

# Public registry: model id -> spec.
DENOISERS: dict[str, DenoiserSpec] = {spec.model_id: spec for spec in _DENOISER_SPECS}
DENOISER_IDS: tuple[str, ...] = tuple(DENOISERS.keys())


def resolve_denoiser(name: str) -> AudioFilter:
    """Resolve a string denoiser id to a concrete :class:`AudioFilter`.

    Parameters
    ----------
    name:
        A registered denoiser id (see :data:`DENOISER_IDS`).

    Returns
    -------
    AudioFilter
        A ready-to-use filter instance for the pipeline inbound chain.

    Raises
    ------
    ValueError
        If ``name`` is not a registered denoiser id.
    RuntimeError
        If the required Krisp model directory (``KRISP_MODELS_DIR``) is unset,
        or the ``krisp-audio`` SDK / license is missing.
    FileNotFoundError
        If the model directory is set but the model ``.kef`` file is absent.
    """
    spec = DENOISERS.get(name)
    if spec is None:
        known = ", ".join(sorted(DENOISERS))
        raise ValueError(f"Unknown denoiser {name!r}. Known denoisers: {known}.")
    return _resolve_krisp(spec)


def _resolve_krisp(spec: DenoiserSpec) -> AudioFilter:
    """Build a :class:`KrispVivaFilter` for a Krisp-family denoiser spec."""
    models_dir = os.getenv(KRISP_MODELS_DIR_ENV)
    if not models_dir:
        raise RuntimeError(
            f"Denoiser {spec.model_id!r} needs the Krisp model directory: set "
            f"{KRISP_MODELS_DIR_ENV} to the folder containing {spec.filename!r}. "
            "Krisp is bring-your-own-license — Patter ships no Krisp SDK, "
            "license, or model files. Install the SDK (pip install "
            "getpatter[krisp]), set KRISP_VIVA_SDK_LICENSE_KEY, and place your "
            "licensed .kef model in that directory."
        )

    model_path = os.path.join(models_dir, spec.filename)

    # Imported lazily so the heavy ``krisp-audio`` import (and its SDK guard)
    # is only paid when a Krisp denoiser is actually selected.
    from getpatter.providers.krisp_filter import KrispVivaFilter

    return KrispVivaFilter(model_path=model_path, session_type=spec.kind)
